### PR TITLE
[web] Line break algorithm using unicode properties

### PR DESCRIPTION
**Pull request by mdebbar**
Switch to using the unicode data instead of checking manually for new lines and whitespace.

Note: this PR uses minimal parts of the unicode data just to cover the main things (e.g. whitespace, NL, BK, LF, CR, etc). There's another PR coming that will implement the full algorithm that'll take advantage of the entire unicode data.

Didn't add any tests because I'm relying on existing tests to verify that the switch to unicode data doesn't break existing behavior.

TODO:
- [ ] Check if there are any performance implications.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text/line_breaker.dart [hunk 1]
```diff
@@ -43,62 +28,117 @@ class LineBreakResult {
   final LineBreakType type;
 }
 
+/// Normalizes properties that behave the same way into one common property.
+LineCharProperty _normalizeLineProperty(LineCharProperty prop) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/line_breaker.dart:32]: Can this substitution be made in the `_packedLineBreakProperties` so we never need to convert at run time?

File: lib/web_ui/lib/src/engine/text/line_breaker.dart [hunk 2]
```diff
@@ -43,62 +28,117 @@ class LineBreakResult {
...
     }
 
-    if (prev == CharCode.space ||
-        prev == CharCode.tab ||
-        prev == CharCode.hyphen) {
+    // At this point, we've handled new lines, and consumed all spaces (if any).
+    // TODO: Use the 2d table now. See https://www.unicode.org/reports/tr14/tr14-22.html#ExampleTable
+
+    // TODO: After using the 2d table, do:
+    // hasSpaces = false;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/line_breaker.dart:139]: Should these todos be github issues?

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/line_breaker.dart:139]: They are coming in the next PR.

File: lib/web_ui/lib/src/engine/text/line_breaker.dart (hunk 1 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/line_breaker.dart:32]: This normalization is part of the algorithm. I want to keep codegen independent.

The other reason is that in some cases we still need to know the original property (pre-normalization). e.g. `rawCurr == LineCharProperty.LF`.

There are a few optimizations that I can think of. I expect them to have marginal perf improvements but nothing drastic.

Some ideas for my future reference:
- Normalize properties during codegen.
- Reduce the amount of binary searches by either caching at runtime, or codegen'ing a hash map for the most common characters, or both.
- For paragraphs that can fit in a single line, we can avoid a whole bunch of computations.

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/text/line_breaker.dart:32]: How about inside `UnicodePropertyLookup.fromPackedData`?

I'm just throwing ideas out. The PR LGTM. This comment is non-blocking.

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/line_breaker.dart:32]: `UnicodePropertyLookup.fromPackedData` is used by both line breaker and word breaker. They have different properties and different rules for normalization.

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/line_breaker.dart:32]: I see your general point though. And I agree we could do some optimizations there.
